transform/date.go

```diff
@@ -0,0 +1,35 @@
+package transform
+
+import (
+	"strings"
+	"time"
+)
+
+const dateInputFormat = "20060102"
+const dateOutputFormat = "2006-01-02"
+
+type date struct {
```

Comment:
Se você não for adicionar nenhum outro tipo em date, pode usar a notação:

```go
type date time.Time
```

Você continua tendo a habilidade de adicionar métodos e fazer marshalling e unmarshalling customizados

Reply:
Eu tive dois problemas tentando fazer dessa maneira.

**Como fazer o `UnmarshalJSON`?**

Com algo assim, eu não consigo altertar o `d` de dentro do método `UnmarshalJSON` aparentemente:

```go
type date time.Time

func (d date) UnmarshalJSON(b []byte) error {
	s := strings.Trim(string(b), `"`)
	if s == "" {
		return nil
	}
	t, err := time.Parse(dateOutputFormat, s)
	if err != nil {
		return err
	}
	d = date(t)
	return nil
}
```

Testes como:

```
d := date{}
d.UnmarshalJSON([]byte("\"1967-06-30\""))
got := time.Time(d)
```

Falham dizendo que ano é 1, o valor “nulo” de quanto crio `date{}`.

**Perco os métodos do `time.Time`**

Tenho que fazer uma conversões `date` para `time.Time` quando preciso acessar os métodos do `time.Time`, por exemplo, no `MarshalJSON`:

```go
func (d date) MarshalJSON() ([]byte, error) {
	t := time.Time(d)
	return []byte(`"` + t.Format(dateOutputFormat) + `"`), nil
}
```

---

O segundo ponto é de boa, convivo com isso se for melhor prática. Mas e o primeiro, como resolver?

Reply:
Ótimos apontamentos, @cuducos ! Segue abaixo dois snippets de código que podem te ajudar (e resolvem os problemas que você citou). Fica a vontade para me chamar no chat e tirar dúvidas: 

* [Exemplo de código com type](https://go.dev/play/p/1uVx7l4XlJQ)
* [Exemplo com embedded type](https://go.dev/play/p/RIdG3GynLhf) (que te dá os métodos de time)

Reply:
Não sabia que rolava fazer `*foo = bar` (o uso do ponteiro do lado de lá da atribuição)… ficou bem melhor a implementação, eu achei a7b06e8 (inclusive nos detalhes de não precisar criar o `null` manualmente no `Marshal`! Obrigado : )